framework/db/SqlToken.php

```diff
@@ -0,0 +1,296 @@
+        }
+        $this->updateCollectionOffsets();
+    }
+
+    /**
+     * Removes a child token at the specified offset.
+     * This method is required by the SPL [[\ArrayAccess]] interface.
+     * It is implicitly called when you use something like `unset($token[$offset])`.
+     * @param int $offset child token offset.
+     */
+    public function offsetUnset($offset)
+    {
+        $offset = $this->calculateOffset($offset);
+        if (isset($this->_children[$offset])) {
+            array_splice($this->_children, $offset, 1);
+        }
+        $this->updateCollectionOffsets();
+    }
+
+    /**
+     * Returns child tokens.
+     * @return SqlToken[] child tokens.
+     */
+    public function getChildren()
+    {
+        return $this->_children;
+    }
+
+    /**
+     * Sets a list of child tokens.
+     * @param SqlToken[] $children child tokens.
+     */
+    public function setChildren($children)
+    {
+        $this->_children = [];
+        foreach ($children as $child) {
+            $child->parent = $this;
+            $this->_children[] = $child;
+        }
+        $this->updateCollectionOffsets();
+    }
+
+    /**
+     * Returns whether the token represents a collection of tokens.
+     * @return bool whether the token represents a collection of tokens.
+     */
+    public function getIsCollection()
+    {
+        return in_array($this->type, [
+            self::TYPE_CODE,
+            self::TYPE_STATEMENT,
+            self::TYPE_PARENTHESIS,
+        ], true);
+    }
+
+    /**
+     * Returns whether the token represents a collection of tokens and has non-zero number of children.
+     * @return bool whether the token has children.
+     */
+    public function getHasChildren()
+    {
+        return $this->getIsCollection() && !empty($this->_children);
+    }
+
+    /**
+     * Returns the SQL code representing the token.
+     * @return string SQL code.
+     */
+    public function getSql()
+    {
+        $code = $this;
+        while ($code->parent !== null) {
+            $code = $code->parent;
+        }
+        return mb_substr($code->content, $this->startOffset, $this->endOffset - $this->startOffset, 'UTF-8');
+    }
+
+    /**
+     * @param string $pattern
+     * @param int $offset
+     * @param int|null $firstMatchIndex
+     * @param int|null $lastMatchIndex
+     * @return bool
+     */
+    public function matches($pattern, $offset = 0, &$firstMatchIndex = null, &$lastMatchIndex = null)
+    {
+        $patternToken = (new \yii\db\sqlite\SqlTokenizer($pattern))->tokenize();
+        if (!$patternToken->getHasChildren()) {
+            return false;
+        }
+
+        $patternToken = $patternToken[0];
+        return $this->tokensMatch($patternToken, $this, $offset, $firstMatchIndex, $lastMatchIndex);
+    }
+
+    /**
+     * Tests the given token to match the specified pattern token.
+     * @param SqlToken $patternToken
```

Comment:
I don't think there's any need for param docs as I covered it in the public method. But anyway take a look at it again after a commit is pushed.